docs/design/object-locking.md

```diff
@@ -0,0 +1,187 @@
+# Object Locking using S3 Object Lock
+
+Object locking feature is built for addressing use cases where
+Write Once Read Many (WORM) model is required. The feature will be built in line
+with AWS specification, any extensions to the specification will be explicitly
+documented. The feature implementation has the goal of meeting
+SEC 17a-4 compliance.
+
+## Requirements
+
+* Bucket needs to have object lock flag set during bucket creation
+* Versioning has to be enabled on the bucket
+* Lock configuration can be written only on a bucket using PUT Object Lock
+  Configuration api that has object lock flag set
+
+PS: AWS specification does not have a way of setting object lock on existing
+buckets cannot have object lock flag set through the api. To enable
+object lock flag on an existing bucket, a migration needs to be performed
+on the bucket using a tool. Objects that were created before the lock is set
```

Comment:
I think this last sentence should be its own note or PS. And it should read ...`before the lock _configuration_ is set...`

Reply:
Done